app/vmalert/manager.go

```diff
@@ -25,52 +25,61 @@ type manager struct {
 	wg     sync.WaitGroup
 	labels map[string]string
 
-	groupsMu sync.RWMutex
-	groups   map[uint64]*rule.Group
+	groupsMu  sync.RWMutex
+	groups    []*rule.Group
+	groupsIds map[uint64]int
```

Comment:
I don't like the pagination implementation complicates the internal manager code, which also requires recreating both objects during `manager.update()`.
Can we isolate the pagination within the web package? During the API call, it can retrieve the group info from a global config file or raw group list, handle the object conversions, etc... This should not slow down the API too much but would make the functions more isolated.

Reply:
It's completely fine to copy groups refs in-memory, sort and apply needed pagination. There is no requirement for any kind of speed or accuracy of pagination for this API.

 But it will keep change small and isolated. Current PR introduces a tone of potential bugs and adds unrelated to primary issue changes.

Reply:
@f41gh7 @Haleygo updated PR